### PR TITLE
Compatibility with new lalsimulation waveform interface

### PR DESCRIPTION
**Pull request by duncan-brown**
This include @sturani's functionality from https://github.com/ligo-cbc/pycbc/pull/1174 and also updates the travis build to use a version of lalsuite that has the new changes.

### CODE REVIEW

**Comment by ahnitz** [comment]
1) @duncan-brown I am re-opening Riccardos original pull request. That can be merged in sequence with this one. Depends on #1174. This ensures that Riccardo gets some credit, as the default method to merge is to first squash the commit history. 

2) I don't believe this address all issues, as recently demonstrated by Bernd's test machine. It appears the TF2 interface was also updated. This will need a corresponding patch to our code. 

https://github.com/lscsoft/lalsuite/blob/master/lalsimulation/src/LALSimInspiralTaylorF2.c#L63

3)In any case, as this is a known break in compatibility, we need to see a run of the test workflow.

**Comment by ahnitz** [comment]
Also, note that when #1224  and #1174 are merged, there should be an email sent to the CBC list. This means PyCBC will be *incompatible* with all prior lalsuite release, and that people must update to use the latest code.

**Comment by spxiwh** [comment]
I'm not making any statement on whether this one or #1174 is the right one to merge, as I haven't even started to catch up yet!

However, in cases like this where there is input from more than one person and one wants to preserve the git history, use "Rebase and merge" instead of the default "Squash and merge" button when merging. Using "Squash and merge" will compress everything into a single commit from a single user. That option is enabled for exactly this reason. Disadvantage to that is that you manage the commit history yourself (So no "Fix whitespace" -> "FIx Whitespace again" -> "Landscape" -> "WS" -> "." -> "Patch" type list of commits)

**Comment by duncan-brown** [comment]
@ahnitz, @spxiwh is correct. We don't *have* to squash and merge, especially if a pull request contains patches from multiple people. That is the way git is supposed to work. Having me submit a request against @sturani's branch is more work for me and him than just pushing the "Rebase and merge" button.

**Comment by duncan-brown** [comment]
@ahnitz I'm going to re-analyze all of O1 using O1 tuning and ER10 tuning, so this will throughly test this. I would like to do it with a release, however. We can always do another point release if there are issues with the LAL code.

This is a sufficiently major change that I think we need a *complete* re-analysis of O1 to test it, as waveform bugs often show up in edge cases.

**Comment by ahnitz** [comment]
@duncan-brown The test workflow is designed specifically for this case and should run quickly. We can't put master in a broken state, so if you are adamant on merging this soon, I ask that a test workflow demonstrate these changes before they are merged to master.

**Comment by duncan-brown** [comment]
I'm putting this on hold for now.

**Comment by cdcapano** [comment]
I just realized I can't test the new SEOBNRv4 until this and/or #1174 are merged to master. To do the sainity checks on the waveform for the SEOBNRv4 review, I used this script from pycbc:
[https://www.lsc-group.phys.uwm.edu/ligovirgo/cbcnote/WaveformsReview/SEOBNRv4SainityCheckScript?action=AttachFile&do=view&target=test_lalsim.py](url)

I can't redo that test on the new v4 until this code is merged, since it uses get_td_waveform in waveform.py. Can this be done soon?

**Comment by duncan-brown** [comment]
@cdcapano I won't get back to this before Wednesday at the earliest. If you want to fix this in the mean time, please go ahead.

**Comment by cdcapano** [comment]
Just a test workflow needs to be run?

**Comment by sturani** [comment]
Hi Collin,

would it help if I edited the test_lalsim.py script in order to call the
new ChooseWform interface?
Cheers,

R

**Comment by cdcapano** [comment]
@sturani Sure, although this patch will need to be merged ASAP anyway, else the pycbc workflow won't be able to use the updated SEOBNRv4. Might be better to just invest the time in getting this ready to merge.

**Comment by cdcapano** [comment]
@duncan-brown @ahnitz @spxiwh If I wanted to run a test workflow, which branch do I need to check out? `duncan-brown:lal_waveform`? Does this include the changes in `sturani:NewChooseWform`? I'm a bit confused as to the differences between this and #1174.

**Comment by ahnitz** [comment]
@cdcapano As far as I can see, there is also some work needed to update the F2 phasing coefficients call that SPAtmplt needs.

**Comment by duncan-brown** [comment]
@cdcapano let me rebase ``duncan-brown:lal_waveform`` to master.

**Comment by sturani** [comment]
Hi All,

I tested the file
pycbc/test/test_lalsim.py
with the newChooseWform patch
and I see nothing wrong with that.
@Collin, the file attached to the wiki you mentioned is not updated with
the one in the git repo, pls update it.

Cheers,

R

**Comment by duncan-brown** [comment]
@cdcapano ok, I think my branch should be ready for you to run a test workflow. I fixed the interface and squashed the commits together to make sure that there are no bad commits

@ahnitz to make everyone's lives easier, *please* can we close https://github.com/ligo-cbc/pycbc/pull/1174 and just use this pull request for the discussion. As far as I can tell the only reason to keep https://github.com/ligo-cbc/pycbc/pull/1174 open is so that @sturani gets credit for the commit, and that can be handed by @cdcapano doing a rebase and merge not a squash and merge on this pull request.

**Comment by duncan-brown** [comment]
OK, @cdcapano over to you. Thanks for offering to test.

**Comment by duncan-brown** [comment]
@cdcapano I started a run here:

https://sugwg-osg.phy.syr.edu/pegasus/u/dbrown/r/336/w?wf_uuid=8ac93189-a672-41ca-b24e-b386f793637c

**Comment by cdcapano** [comment]
So I shouldn't run a test workflow now?

**Comment by duncan-brown** [comment]
@cdcapano no, please go ahead and also test.

**Comment by cdcapano** [comment]
@duncan-brown Ok, test run is running here: https://atlas6/pegasus/u/cdcapano/r/41/w?wf_uuid=e3a49851-01e1-4199-ae7b-94483fe3ed7a

**Comment by cdcapano** [comment]
@duncan-brown @ahnitz My test workflow got through the inspiral jobs, but it's failing on others. Not sure why. Here is the ouput of `debug`: [link](https://www.atlas.aei.uni-hannover.de/~cdcapano/LSC/scratch/lal_waveform_errors.txt).

**Comment by duncan-brown** [comment]
@cdcapano looking at the errors in turn:
```
AttributeError: 'module' object has no attribute 'LineHTMLTooltip'
```
seems like a bad mpld3 install. Can you try
```
pip uninstall mpld3
pip install --no-cache pycbc-mpld3
```

Then
```
./injections-BNS01_INJ-HL_ID8: error while loading shared libraries: libhdf5.so.7: cannot open shared object file: No such file or directory
```
seems to be a problem with the lalapps build. It's not linking in HDF5. I suggest you follow the instructions [http://ligo-cbc.github.io/pycbc/latest/html/install_lalsuite.html#building-and-installing-into-your-virtual-environment](here) and do a static build of the lalapps code. I have had no end of problems with dynamic lalapps executables.

Lastly
```
  File "./combine_statmap-FULL_DATA_FULL_CUMULATIVE_CAT_12H-H1L1_ID17", line 4, in <module>
    __import__('pkg_resources').run_script('PyCBC==4094a8', 'pycbc_combine_statmap')
  File "/home/cdcapano/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 743, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/cdcapano/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1498, in run_script
    exec(code, namespace, namespace)
  File "/home/cdcapano/.virtualenvs/lal_waveform/lib/python2.7/site-packages/PyCBC-4094a8-py2.7.egg/EGG-INFO/scripts/pycbc_combine_statmap", line 84, in <module>
    dset = f['foreground/%s' % key][:][cidx]
IndexError: arrays used as indices must be of integer (or boolean) type
```
looks like a genuine error with statmap that @ahnitz should look at. Perhaps an unhandled empty array>

**Comment by ahnitz** [comment]
@cdcapano On your hdf5 error. Did you you rebuild lalsuite since the upgrade? I only get the error when I forget and source an old install that is lying around. You *are not* required to build a static version. Also, please record what version extatly of mpld3 your installation is trying to use before you change anything. It is critical that we actually understand these errors and not just sweep them under the rug.

File: pycbc/waveform/spa_tmplt.py [hunk 1]
```diff
@@ -183,15 +174,20 @@ def spa_tmplt(**kwds):
 
     amp_factor = spa_amplitude_factor(mass1=mass1, mass2=mass2) / distance
 
+    lal_pars = lal.CreateDict()
+    if phase_order != -1:
+        lalsimulation.SimInspiralWaveformParamsInsertPNPhaseOrder(
+            lal_pars,phase_order)
```
Comment (spxiwh) [inline pycbc/waveform/spa_tmplt.py]: For things like this, we can just fix directly. I will try to do this ...

**Comment by duncan-brown** [comment]
@spxiwh hold your horses, I'm trying to test this branch!

**Comment by ahnitz** [comment]
Collin, I looked at your mpld3 error a bit more, and I'm mystified. The
mpld3 in your virtualenv has this class right there in the file. Please
don't change anything. It is worth looking into this a bit more to figure
out what the heck is going on. If at all possible though, can you allow
other user to source your environment? It is all chmodded such that we
can't.

**Comment by cdcapano** [comment]
@duncan-brown @ahnitz I fixed the issues with my virtual environment that caused the errors I got in the last test workflow and started a new one. I also pulled the last update that Duncan pushed. Will let you know when the workflow is finished.

File: tools/install_travis.sh [hunk 2]
```diff
@@ -57,7 +57,7 @@ else
     git clone -q https://github.com/lscsoft/lalsuite.git
     cd lalsuite
     # see PR #1067 for the choice of commit
```
Comment (titodalcanton) [inline tools/install_travis.sh]: Can somebody update or remove this comment, since the git hash has been changed?

**Comment by cdcapano** [comment]
There seems to have been some work done at Syracuse that hasn't been communicated to the rest of the group, so I'm posting here what I know to at least get some documentation out there. First, my test run failed due to apparently unrelated problems with the empty trigger files (this has been discussed in the Pycbc Code channel on pycbc mattermost). However, @duncan-brown was able to get a test run to complete by using this patch with an older version (1.5.7) of pycbc. Here are his results before this patch:
https://sugwg-jobs.phy.syr.edu/~dbrown/ER10/review/o1-analysis-3-er10-v1.5.7/
and after:
https://sugwg-jobs.phy.syr.edu/~dbrown/ER10/review/o1-analysis-3-er10-lal-wf/

There are some differences in the found/missed injections that, as I understand, @cmbiwer is looking into. Is that right @duncan-brown and @cmbiwer?

In any event, @ahnitz and I briefly looked at Duncan's results, and we quickly discussed them on today's (11/23) pycbc call. @spxiwh pointed out that, depending on the approximant, the definition of orbital phase has changed, (this is definitely true for SEOBNR).  Thus, even if you use exactly the same injection xml file, the actual waveforms generated will have a different phase, which can lead to a change in SNR by O(1), depending on the noise realization. @spxiwh suggested plotting the actual waveforms created before/after the change of an injection.

**Comment by duncan-brown** [comment]
@cdcapano these are the runs that I pinged you about in the PyCBC code chat on Nov 21. There do seem to be some large differences in the sensitivity (see the attached plots).

h1l1-plot_sensitivity_all_spin_allinj-1128299417-1083600

h1l1-plot_sensitivity_all_spin_allinj-1128299417-1083600-1

The only difference here is the LAL version. @spxiwh are there additional changes to SEOBNRv4 that explain this? It seems to large a difference to just be waveform phase.

**Comment by ahnitz** [comment]
Duncan, I'm not sure I see what is larger than expect variation for randomizing the phases (or conversely think randomizing the noise). The error bars for the red line are pretty large.

**Comment by duncan-brown** [comment]
@ahnitz the error bars are large, but the scales on the two plots are different. Take an IFAR of 10^2. One plot has (eyeballing) 1200(+100)(-200) Mpc, the other has 1050(+50)(-50) Mpc. That's quite a large difference. I think this needs some following up.

**Comment by spxiwh** [comment]
Most of the difference in the plots Duncan links seems to be down to a single, large-mass IMBH injection, that is seen with FAR 0 in the old case and missed completely in the new. That *could* happen because of the phase difference.

**Comment by sturani** [comment]
Hi All,

WARNING:
git commit 62e5be3c1eef58c0234ca034939ae7e96da2f767
to master on Nov 21st has indeed changed SEOBNRv4 as per
https://www.lsc-group.phys.uwm.edu/ligovirgo/cbcnote/WaveformsReview/SEOBNRv4/update
so if these runs picked this version they have been using a different
SEOBNRv4 than before.
Cheers,

Riccardo

**Comment by duncan-brown** [comment]
@cdcapano I'm happy to merge.
